File: src/lib/songLibrary.ts
import { mockSongs } from "./mockSongs";
import { storage } from "./storage";

export interface SongData {
  code: string;
  artist: string;
  title: string;
  verse: string;
  filename: string;
}

export class SongLibrary {
  private songs: Map<string, SongData> = new Map();
  private mediaPath: string;
  private initialized: boolean = false;

  constructor() {
    this.mediaPath = "/media/karaoke";
    this.loadDatabase().catch(console.error);
  }

  private async loadDatabase() {
    if (this.initialized) return;

    try {
      // Try to load from db.ini first
      const content = await storage.readFile("db.ini");
      if (content) {
        const lines = content.split("\n");
        lines.forEach((line) => {
          const [code, artist, title, filename] = line
            .split("|")
            .map((s) => s.trim());
          if (code && artist && title && filename) {
            this.songs.set(code, {
              code,
              artist,
              title,
              verse: "",
              filename,
            });
          }
        });
        this.initialized = true;
        return;
      }
    } catch (error) {
      console.error("Error loading database:", error);
    }

    // Fallback to mock songs if db.ini not found or error
    mockSongs.forEach((song) => {
      this.songs.set(song.code, song);
    });
    this.initialized = true;
  }

  findSong(code: string): SongData | null {
    // If not initialized, use mock songs temporarily
    if (!this.initialized) {
      const mockSong = mockSongs.find((s) => s.code === code);
      return mockSong || null;
    }
    return this.songs.get(code) || null;
  }

  getSongUrl(filename: string): string {
    if (import.meta.env.DEV) {
      return "https://storage.googleapis.com/gtv-videos-bucket/sample/BigBuckBunny.mp4";
    }
    return `${this.mediaPath}/${filename}`;
  }
}

export const songLibrary = new SongLibrary();

File: lib/services/song_library.dart
import 'dart:io';
import 'package:path_provider/path_provider.dart';
import '../models/song.dart';

class SongLibrary {
  static final SongLibrary _instance = SongLibrary._internal();
  factory SongLibrary() => _instance;
  
  SongLibrary._internal();

  final Map<String, Song> _songs = {};
  bool _initialized = false;
  String _mediaPath = '/storage/emulated/0/Download/karaoke';

  Future<void> initialize() async {
    if (_initialized) return;

    try {
      final directory = await getExternalStorageDirectory();
      if (directory != null) {
        final dbFile = File('${directory.path}/db.ini');
        if (await dbFile.exists()) {
          final lines = await dbFile.readAsLines();
          for (var line in lines) {
            final parts = line.split('|');
            if (parts.length >= 4) {
              final song = Song(
                code: parts[0].trim(),
                artist: parts[1].trim(),
                title: parts[2].trim(),
                videoUrl: '${_mediaPath}/${parts[3].trim()}',
              );
              _songs[song.code] = song;
            }
          }
        }
      }
    } catch (e) {
      print('Error loading song database: $e');
      // Carregar algumas músicas de exemplo para teste
      _loadMockSongs();
    }

    _initialized = true;
  }

  void _loadMockSongs() {
    final mockSongs = [
      Song(
        code: '12345',
        artist: 'Test Artist',
        title: 'Test Song',
        videoUrl: 'https://example.com/test.mp4',
      ),
      // Adicione mais músicas de exemplo conforme necessário
    ];

    for (var song in mockSongs) {
      _songs[song.code] = song;
    }
  }

  Song? findSong(String code) {
    return _songs[code];
  }

  List<Song> searchSongs(String query) {
    query = query.toLowerCase();
    return _songs.values.where((song) {
      return song.title.toLowerCase().contains(query) ||
             song.artist.toLowerCase().contains(query) ||
             song.code.toLowerCase().contains(query);
    }).toList();
  }
}